mmdet/models/dense_heads/ddod_head.py

```diff
@@ -0,0 +1,641 @@
+            label_channels (int): Channel of label.
+            unmap_outputs (bool): Whether to map outputs back to the original
+                set of anchors.
+            is_cls_assigner (bool): Classification or regression.
+
+        Returns:
+            tuple: N is the number of total anchors in the image.
+                - labels (Tensor): Labels of all anchors in the image with \
+                    shape (N,).
+                - label_weights (Tensor): Label weights of all anchor in the \
+                    image with shape (N,).
+                - bbox_targets (Tensor): BBox targets of all anchors in the \
+                    image with shape (N, 4).
+                - bbox_weights (Tensor): BBox weights of all anchors in the \
+                    image with shape (N, 4)
+                - pos_inds (Tensor): Indices of positive anchor with shape \
+                    (num_pos,).
+                - neg_inds (Tensor): Indices of negative anchor with shape \
+                    (num_neg,).
+        """
+        inside_flags = anchor_inside_flags(flat_anchors, valid_flags,
+                                           img_meta['img_shape'][:2],
+                                           self.train_cfg.allowed_border)
+        if not inside_flags.any():
+            return (None, ) * 7
+        # assign gt and sample anchors
+        anchors = flat_anchors[inside_flags, :]
+
+        num_level_anchors_inside = self.get_num_level_anchors_inside(
+            num_level_anchors, inside_flags)
+        bbox_preds_valid = bbox_preds[inside_flags, :]
+        cls_scores_valid = cls_scores[inside_flags, :]
+
+        assigner = self.cls_assigner if is_cls_assigner else self.reg_assigner
+
+        # decode prediction out of assigner
+        bbox_preds_valid = self.bbox_coder.decode(anchors, bbox_preds_valid)
+        assign_result = assigner.assign(anchors, num_level_anchors_inside,
+                                        cls_scores_valid, bbox_preds_valid,
+                                        gt_bboxes, gt_bboxes_ignore, gt_labels)
+        sampling_result = self.sampler.sample(assign_result, anchors,
+                                              gt_bboxes)
+
+        num_valid_anchors = anchors.shape[0]
+        bbox_targets = torch.zeros_like(anchors)
+        bbox_weights = torch.zeros_like(anchors)
+        labels = anchors.new_full((num_valid_anchors, ),
+                                  self.num_classes,
+                                  dtype=torch.long)
+        label_weights = anchors.new_zeros(num_valid_anchors, dtype=torch.float)
+
+        pos_inds = sampling_result.pos_inds
+        neg_inds = sampling_result.neg_inds
+        if len(pos_inds) > 0:
+            if hasattr(self, 'bbox_coder'):
+                pos_bbox_targets = self.bbox_coder.encode(
+                    sampling_result.pos_bboxes, sampling_result.pos_gt_bboxes)
+            else:
+                # used in VFNetHead
```

Comment:
VFNet does not inherit this class, is this part necessary?

Reply:
Yes. It's necessary for the possible situation.